Improve the error by showing wich object entity was being used when t…

odnanref opened the pull request:
…he method was detected missing

I had a error that didn't show which class didn't have the method, this litle hack allowed me to have that information.
The current error shows withouth the entity to which the method is missing.
 request.CRITICAL: Uncaught PHP Exception BadMethodCallException: " Undefined method 'getAll'. The method name must start with either findBy or findOneBy!" at /home/andref/workspace/moneytrak/mt2/vendor/doctrine/orm/lib/Doctrine/ORM/EntityRepository.php line 196 {"exception":"[object](BadMethodCallException:  Undefined method 'getAll'. The method name must start with either findBy or findOneBy! at /home/andref/workspace/moneytrak/mt2/vendor/doctrine/orm/lib/Doctrine/ORM/EntityRepository.php:196)"} []

This change improves the error shown:
 request.CRITICAL: Uncaught PHP Exception BadMethodCallException: "In Far\MT\AccountBundle\Entity\Holder Undefined method 'getAll'. The method name must start with either findBy or findOneBy!" at /home/andref/workspace/moneytrak/mt2/vendor/doctrine/orm/lib/Doctrine/ORM/EntityRepository.php line 196 {"exception":"[object](BadMethodCallException: In Far\MT\AccountBundle\Entity\Holder Undefined method 'getAll'. The method name must start with either findBy or findOneBy! at /home/andref/workspace/moneytrak/mt2/vendor/doctrine/orm/lib/Doctrine/ORM/EntityRepository.php:196)"} []

Now I have the Entity to which the method is missing.
(BadMethodCallException: In Far\MT\AccountBundle\Entity\Holder

odnanref committed:
Improve the error by showing wich object entity was being used when the method was detected missing

I had a error that didn't show which class didn't have the method, this litle hack allowed me to have that information.
The current error shows withouth the entity to which the method is missing.
 request.CRITICAL: Uncaught PHP Exception BadMethodCallException: " Undefined method 'getAll'. The method name must start with either findBy or findOneBy!" at /home/andref/workspace/moneytrak/mt2/vendor/doctrine/orm/lib/Doctrine/ORM/EntityRepository.php line 196 {"exception":"[object] (BadMethodCallException:  Undefined method 'getAll'. The method name must start with either findBy or findOneBy! at /home/andref/workspace/moneytrak/mt2/vendor/doctrine/orm/lib/Doctrine/ORM/EntityRepository.php:196)"} []

This change improves the error shown:
 request.CRITICAL: Uncaught PHP Exception BadMethodCallException: "In Far\MT\AccountBundle\Entity\Holder Undefined method 'getAll'. The method name must start with either findBy or findOneBy!" at /home/andref/workspace/moneytrak/mt2/vendor/doctrine/orm/lib/Doctrine/ORM/EntityRepository.php line 196 {"exception":"[object] (BadMethodCallException: In Far\\MT\\AccountBundle\\Entity\\Holder Undefined method 'getAll'. The method name must start with either findBy or findOneBy! at /home/andref/workspace/moneytrak/mt2/vendor/doctrine/orm/lib/Doctrine/ORM/EntityRepository.php:196)"} []

Now I have the Entity to which the method is missing.
(BadMethodCallException: In Far\\MT\\AccountBundle\\Entity\\Holder

odnanref committed:
better description of error

Rewritten so it would be understandable that the mistake is in the Repository

linaori commented on lib/Doctrine/ORM/EntityRepository.php:
```diff
@@ -225,7 +225,8 @@ public function __call($method, $arguments)
-                    "Undefined method '$method'. The method name must start with ".
+                    "In Repository for Entity " . $this->_entityName . 
+                    " Undefined method '$method'. The method name must start with ".
                     "either findBy or findOneBy!"
```
I would like something more in this direction, as it's more verbose:

``` php
throw new \BadMethodCallException(
    "Undefined method called: '".get_class($this)."::{$method}()'. The ".
    "repository method name must start with either 'findBy' or 'findOneBy'."
);
```

This would produce "Undefined method called: 'Foo\Bar\BazRepository::getAll()'. The repository method name must start with either 'findBy' or 'findOneBy'."

_Note that I left out the exclamation mark on purpose_

odnanref commented:
get_class($this) would get the current class name and not the entity that was related to the error.

odnanref committed:
simplified the writing

"If absent the repository" believe this is better then just saying The Repository has it points to a missing findBy or findOneBy in case there was no other method.

odnanref commented:
Don't understand why the build failed.

Ocramius commented:
@odnanref restarted it

odnanref commented:
thanks

Ocramius commented:
@odnanref could you add tests for this change? As a suggestion, you may move this to a custom exception (extending `BadMethodCallException`) as per http://rosstuck.com/formatting-exception-messages/

odnanref commented:
I'll look into it. UnknownEntityMethod Exception then?

odnanref committed:
throw this when you dont find a method in the repository class

odnanref committed:
throw the UnknownEntityRepositoryMethod that inherits BadMethodException

odnanref committed:
test for the implementation , didnt test the test locally

odnanref committed:
fix syntax error

odnanref commented:
@Ocramius Look at what was done and say what you think.
